### PR TITLE
Add support for IPC4 loadable modules and smart_amp_test module

### PR DESCRIPTION
**Pull request by aiChaoSONG**
All IPC4 loadable module will require base module config + base module config extension style blob to be initialized. the IPC4 smart_amp_test module is implemented with the same style init blob.  Add the support in the kernel for them.

Patches have been verified with IADK  dummy smart amp loadable module, and SOF smart_amp_test module here: https://github.com/thesofproject/sof/pull/6751

### CODE REVIEW

File: sound/soc/sof/ipc4-topology.h [hunk 1]
```diff
@@ -349,4 +349,40 @@ struct sof_ipc4_process {
...
+ * @source_pin_fmt: pin format array for source pins
+ * @priv: module specific parameters follow this structure
+ */
+struct sof_ipc4_base_module_cfg_ext {
+	u16 num_sink_pin_fmts;
+	u16 num_source_pin_fmts;
+	u8 reserved[8];
+	u32 priv_param_len;
+	struct sof_ipc4_pin_format sink_pin_fmt[0];
+	struct sof_ipc4_pin_format source_pin_fmt[0];
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.h:384]: use DECLARE_FLEX_ARRAY() ?

File: sound/soc/sof/sof-audio.h [hunk 2]
```diff
@@ -420,6 +424,15 @@ struct snd_sof_widget {
...
 	bool payload_with_output_fmt;
 
+	/*
+	 * There are two styles of SOF module instance initialization blob:
+	 *   - base module config + module specific extension
+	 *   - base module config + base module config extension
+	 *
+	 * This flag is used to tell which initialization blob style a widget is using.
+	 */
+	int init_blob_style;
```
Comment (ranj063) [inline sound/soc/sof/sof-audio.h:434]: this is IPC4 specific isnt it? Does it need to go in struct snd_sof_widget? Can we find a better place?

File: sound/soc/sof/ipc4-topology.c [hunk 3]
```diff
@@ -1749,6 +1750,15 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 	}
 
 	memcpy(cfg, data, control_data->data->size);
+
+	/* Update pipeline params from output format */
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c:1754]: does this conflict with @libinyang 's PR?

File: sound/soc/sof/ipc4-topology.h [hunk 4]
```diff
@@ -349,4 +349,40 @@ struct sof_ipc4_process {
 	struct sof_ipc4_msg msg;
 };
 
+/**
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.h:352]: are we already using this type "base module config + module specific extension"?

File: sound/soc/sof/ipc4-topology.c [hunk 5]
```diff
@@ -1600,16 +1601,19 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
...
+	int data_size;
 	void *data;
 	int ret, i;
 
 	available_fmt->ref_audio_fmt = &available_fmt->base_config->audio_fmt;
 
 	/*
-	 * Output format is optional for process modules.
-	 * Process modules setup the output format based on audio format tokens in topology.
+	 * Output format is only valid and optional for modules with base module config +
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c:1681]: i'd switch around the optional and only valid bits. "Output format is optional and only valid ...." and the check below as well

File: sound/soc/sof/ipc4-topology.c [hunk 6]
```diff
@@ -1650,12 +1655,31 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
...
+
+		base_ext = (struct sof_ipc4_base_module_cfg_ext *)data;
+		size = sizeof(struct sof_ipc4_base_module_cfg_ext) +
+					base_ext->priv_param_len +
+					(base_ext->num_sink_pin_fmts +
+					base_ext->num_source_pin_fmts) *
+					sizeof(struct sof_ipc4_pin_format);
+		if (data_size != size) {
+			dev_err(sdev->dev, "Insufficient data for widget %s initialization\n",
+				swidget->widget->name);
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c:1747]: should this check be done on line 1660? What if there are no kcontrols for the widget?

File: sound/soc/sof/ipc4-topology.c [hunk 7]
```diff
@@ -168,6 +168,75 @@ static void sof_ipc4_dbg_audio_format(struct device *dev,
 	}
 }
 
+static const struct sof_ipc4_audio_format *
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c:171]: typo in commit message " For modules using..."

File: sound/soc/sof/ipc4-topology.c [hunk 8]
```diff
@@ -168,6 +168,75 @@ static void sof_ipc4_dbg_audio_format(struct device *dev,
...
+	const struct snd_kcontrol_new *kc;
+	struct snd_sof_control *scontrol;
+	struct soc_bytes_ext *sbe;
+	int pin_start, pin_end;
+	int i;
+
+	if (swidget->init_blob_style == SOF_INIT_BLOB_BASE_WITH_MOD_EXT) {
+		struct sof_ipc4_base_module_cfg *base_config = swidget->private;
+
+		if (pin_type == SOF_PIN_TYPE_SOURCE && swidget->payload_with_output_fmt) {
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c:188]: when will this be true? If I see the call below, pin_type is always SOF_PIN_TYPE_SINK?

File: sound/soc/sof/ipc4-topology.c [hunk 9]
```diff
@@ -168,6 +168,75 @@ static void sof_ipc4_dbg_audio_format(struct device *dev,
...
+
+	if (!base_ext)
+		return NULL;
+
+	pin_fmt = base_ext->sink_pin_fmt;
+	if (pin_type == SOF_PIN_TYPE_SINK) {
+		pin_start = 0;
+		pin_end = base_ext->num_sink_pin_fmts;
+	} else {
+		pin_start = base_ext->num_sink_pin_fmts;
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c:228]: why do we need the else? same as the comment above pin_type will not be different at all

File: sound/soc/sof/sof-audio.h (hunk 2 above)
Comment (aiChaoSONG) [reply sound/soc/sof/sof-audio.h:434]: snd_sof_widget is shared between IPC3 and IPC4, I don't think we have a IPC4 specific struct which is a good fit for this.

we can later add snd_sof_ipc3_widget/snd_sof_ipc4_widget if necessary.

File: sound/soc/sof/ipc4-topology.c (hunk 3 above)
Comment (aiChaoSONG) [reply sound/soc/sof/ipc4-topology.c:1754]: Yeah, we do the same thing to update pipeline params, we have synced, and I will merge some of his code in this PR, then we can close https://github.com/thesofproject/linux/pull/4097

File: sound/soc/sof/ipc4-topology.h (hunk 4 above)
Comment (aiChaoSONG) [reply sound/soc/sof/ipc4-topology.h:352]: yes, all bese fw module use this type, for example: 
copier:
```
struct ipc4_copier_module_cfg {
	struct ipc4_base_module_cfg base;

	struct ipc4_audio_format out_fmt;
	uint32_t copier_feature_mask;
	struct ipc4_copier_gateway_cfg  gtw_cfg;
} __attribute__((packed, aligned(4)));
```

peakvol
```
struct ipc4_peak_volume_module_cfg {
	struct ipc4_base_module_cfg base_cfg;
	struct ipc4_peak_volume_config config[0];
} __packed __aligned(8);
```

module specific extension is optional, mixin/mixout don't have module specific extension.
https://github.com/thesofproject/sof/blob/73f6171bb3c8c5b1d0963883014508022445b997/src/audio/mixin_mixout/mixin_mixout.c#L157

File: sound/soc/sof/ipc4-topology.c (hunk 8 above)
Comment (aiChaoSONG) [reply sound/soc/sof/ipc4-topology.c:188]: in updating pipeline params (process module prepare ops), we need to get the output pin format, and set the output format to pipeline param, the pin_type will be SOF_PIN_TYPE_SOURCE there.

File: sound/soc/sof/ipc4-topology.c (hunk 9 above)
Comment (aiChaoSONG) [reply sound/soc/sof/ipc4-topology.c:228]: the function supports to get any input/output pin format, in setting copier sink, we need to get the input format of the widget that connected to copier, in updating pipeline params, we need to get the output format of the widget, and update the format to pipeline params.

File: sound/soc/sof/ipc4-topology.h (hunk 1 above)
Comment (aiChaoSONG) [reply sound/soc/sof/ipc4-topology.h:384]: The init blob comes from topology bytes kcontrol, we will use this structure to decode the binary blob, we don't need to alloc anything, I don't think FLEX_ARRAY is good to be used for this case.

Comment (ranj063) [reply sound/soc/sof/ipc4-topology.h:384]: @aiChaoSONG I dont think it makes sense to have 2 flexible arrays in a struct. Please combine them into one, pin_fmts[]

File: sound/soc/sof/ipc4-topology.h (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.h:352]: @aiChaoSONG can youplease add this to the commit message saying that support for the first type is already implemented

File: sound/soc/sof/sof-audio.h (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.h:434]: How about  struct sof_ipc4_process? We know that the base_fw modules already use module specific extension.

File: sound/soc/sof/ipc4-topology.h [hunk 10]
```diff
@@ -349,4 +349,40 @@ struct sof_ipc4_process {
...
+ * struct sof_ipc4_pin_format - widget's pin format
+ * @pin_index: pin index for the widget
+ * @ibs: input buffer size for input pin
+ * @obs: output buffer size for output pin
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.h:356]: how do we know if the pin is input or output?

File: sound/soc/sof/ipc4-topology.h [hunk 11]
```diff
@@ -349,4 +349,40 @@ struct sof_ipc4_process {
 	struct sof_ipc4_msg msg;
 };
 
+/**
+ * struct sof_ipc4_pin_format - widget's pin format
+ * @pin_index: pin index for the widget
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.h:354]: is the pin_index relative to a direction, i.e. you can have both pin_index=1 for input and pin_index=1 for output

File: sound/soc/sof/ipc4-topology.h (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.h:384]: indeed this is heavily frowned upon. I don't understand what this is about quite frankly.

File: include/uapi/sound/sof/tokens.h [hunk 12]
```diff
@@ -98,6 +98,7 @@
 #define SOF_TKN_COMP_SINK_PIN_BINDING_WNAME	413
 #define SOF_TKN_COMP_SRC_PIN_BINDING_WNAME	414
 #define SOF_TKN_COMP_PAYLOAD_WITH_OUTPUT_FMT	415
+#define SOF_TKN_COMP_INIT_BLOB_STYLE		416
```
Comment (plbossart) [inline include/uapi/sound/sof/tokens.h:101]: what are the values for this token?

File: sound/soc/sof/sof-audio.h [hunk 13]
```diff
@@ -34,6 +34,10 @@
 #define SOF_PIN_TYPE_SINK	0
 #define SOF_PIN_TYPE_SOURCE	1
 
+/* Widget initialization blob style */
+#define SOF_INIT_BLOB_BASE_WITH_MOD_EXT	0
+#define SOF_INIT_BLOB_BASE_WITH_BASE_EXT	1
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.h:39]: well the values should be in the tokens.h definition, no? This defines the UAPI, it should be clear.
I am afraid we have similar issues with the pin type. Gah.

File: sound/soc/sof/sof-audio.h (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.h:434]: payload_with_output_fmt is already IPC4-specific @ranj063, looks like another poorly defined partitioning.

File: sound/soc/sof/ipc4-topology.c [hunk 14]
```diff
@@ -1650,12 +1655,31 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
...
 		break;
 	}
 
+	/* Make sure we have enough data for base module extension structure */
+	if (swidget->init_blob_style == SOF_INIT_BLOB_BASE_WITH_BASE_EXT) {
+		struct sof_ipc4_base_module_cfg_ext *base_ext;
+		int size;
+
+		base_ext = (struct sof_ipc4_base_module_cfg_ext *)data;
+		size = sizeof(struct sof_ipc4_base_module_cfg_ext) +
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:1740]: Can you really do a sizeof() on this:

````c
struct sof_ipc4_base_module_cfg_ext {
	u16 num_sink_pin_fmts;
	u16 num_source_pin_fmts;
	u8 reserved[8];
	u32 priv_param_len;
	struct sof_ipc4_pin_format sink_pin_fmt[0];
	struct sof_ipc4_pin_format source_pin_fmt[0];
````

This is hairy.

File: sound/soc/sof/ipc4-topology.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c:1747]: what if there are more than one kcontrols for the widget?

File: sound/soc/sof/ipc4-topology.c (hunk 7 above)
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:171]: commit message: I have idea what this means

"This patch will set copier sink from the connected pin format."

The connected pin format can only be used to set a sink pin format, no?

File: sound/soc/sof/ipc4-topology.c [hunk 15]
```diff
@@ -168,6 +168,75 @@ static void sof_ipc4_dbg_audio_format(struct device *dev,
 	}
 }
 
+static const struct sof_ipc4_audio_format *
+sof_ipc4_get_widget_pin_format(struct snd_sof_widget *swidget, bool pin_type,
+			       int pin_id)
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:173]: one line

File: sound/soc/sof/ipc4-topology.c [hunk 16]
```diff
@@ -168,6 +168,75 @@ static void sof_ipc4_dbg_audio_format(struct device *dev,
...
+		scontrol = sbe->dobj.private;
+		if (!scontrol)
+			continue;
+		control_data = scontrol->ipc_control_data;
+		if (control_data->data->blob_type != SOF_IPC4_MOD_INIT_INSTANCE)
+			continue;
+
+		base_ext = (struct sof_ipc4_base_module_cfg_ext *)control_data->data->data;
+		break;
+	}
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:218]: this is spaghetti code to me. Just can't figure out how safe/reliable this is.

File: sound/soc/sof/ipc4-topology.c [hunk 17]
```diff
@@ -168,6 +168,75 @@ static void sof_ipc4_dbg_audio_format(struct device *dev,
...
+	if (pin_type == SOF_PIN_TYPE_SINK) {
+		pin_start = 0;
+		pin_end = base_ext->num_sink_pin_fmts;
+	} else {
+		pin_start = base_ext->num_sink_pin_fmts;
+		pin_end = base_ext->num_sink_pin_fmts + base_ext->num_source_pin_fmts;
+	}
+
+	for (i = pin_start; i < pin_end; i++) {
+		if (pin_fmt[i].pin_index == pin_id)
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:233]: you've got to explain how the pins are ordered or numbered. This is super hard to review.

File: sound/soc/sof/ipc4-topology.c [hunk 18]
```diff
@@ -1749,6 +1750,15 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
...
 
 	memcpy(cfg, data, control_data->data->size);
+
+	/* Update pipeline params from output format */
+	out_fmt = sof_ipc4_get_widget_pin_format(swidget, SOF_PIN_TYPE_SOURCE, 0);
+	/* Currently, only smart amp module needs to update channels */
+	hw_param_interval(pipeline_params, SNDRV_PCM_HW_PARAM_CHANNELS)->min =
+		SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(out_fmt->fmt_cfg);
+	hw_param_interval(pipeline_params, SNDRV_PCM_HW_PARAM_CHANNELS)->max =
+		SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(out_fmt->fmt_cfg);
```
Comment (plbossart) [inline sound/soc/sof/ipc4-topology.c:1760]: I don't understand what this is about.

a) does a pipeline have format parameters? If not, then the commit message is just wrong.
b) can avoid the repetition of hw_param_interval(pipeline_params, SNDRV_PCM_HW_PARAM_CHANNELS)-> and use a local variable to make the code more readable?

**Review by plbossart — CHANGES_REQUESTED** [review]
we're quite far from a mergeable PR @aiChaoSONG, please see comments below.

File: sound/soc/sof/sof-audio.h (hunk 2 above)
Comment (bardliao) [reply sound/soc/sof/sof-audio.h:434]: > payload_with_output_fmt is already IPC4-specific @ranj063, looks like another poorly defined partitioning.

Maybe fix both of them?

File: sound/soc/sof/ipc4-topology.c [hunk 19]
```diff
@@ -1650,12 +1655,31 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 		control_data = scontrol->ipc_control_data;
 		if (control_data->data->blob_type != SOF_IPC4_MOD_INIT_INSTANCE)
 			continue;
-
+		data_size = control_data->data->size;
```
Comment (bardliao) [inline sound/soc/sof/ipc4-topology.c:1728]: What if data_size = 0? Or there is no SOF_IPC4_MOD_INIT_INSTANCE blob type in the kcontrol?

File: sound/soc/sof/ipc4-topology.c (hunk 7 above)
Comment (aiChaoSONG) [reply sound/soc/sof/ipc4-topology.c:171]: The already merged PR name the function as `sof_ipc4_set_copier_sink_format` to align with firmware naming, it is actually setting the copier source(output) pin format from the connected sink(input) pin format of the downstream widget.

I will update the commit message to: set copier source pin from connected sink pin format

File: sound/soc/sof/sof-audio.h (hunk 2 above)
Comment (libinyang) [reply sound/soc/sof/sof-audio.h:434]: > payload_with_output_fmt is already IPC4-specific @ranj063, looks like another poorly defined partitioning.

@ujfalusi raised this concern about payload_with_output_fmt before. But we don't have ipc4 related structure at that time, so we put them here temporarily. I think we do need a IPC4 specific place to put ipc4 stuff. If this is only used by process module, struct sof_ipc4_process is good.

File: sound/soc/sof/ipc4-topology.h (hunk 1 above)
Comment (libinyang) [reply sound/soc/sof/ipc4-topology.h:384]: > @aiChaoSONG I dont think it makes sense to have 2 flexible arrays in a struct. Please combine them into one, pin_fmts[]

Actually there are three: ``` struct sof_ipc4_pin_format sink_pin_fmt[0]; struct sof_ipc4_pin_format source_pin_fmt[0]; u8 priv[0];``` Maybe the easiest way is just to use ```void *priv_data[0];``` or ```u8 priv[0]``` directly.

Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.h:384]: This does not compute for me either.
```c
	struct sof_ipc4_pin_format sink_pin_fmt[0];
	struct sof_ipc4_pin_format source_pin_fmt[0];
	u8 priv[0];
```
does not make any sense at all. Where priv[0] is pointing to? They all point to the same offset as all of them have size of 0.
at best you have `data[]` at the end with a comment that from there you have num_sink_pin_fmts x sof_ipc4_pin_format followed by num_source_pin_fmts x sof_ipc4_pin_format then priv_param_len number of bytes of some data.

**Comment by aiChaoSONG** [comment]
a better alternative is #4121, close this one.

**Pull request closed without merging**